### PR TITLE
15447 Register input service refactor

### PR DESCRIPTION
**Pull request by GeorgeChatzigiannis**
Ready for review

### CODE REVIEW

File: pupil-spa/src/app/services/question/question.service.ts [hunk 1]
```diff
@@ -14,7 +14,8 @@ export class QuestionService {
   protected config: Config;
 
   constructor(protected storageService: StorageService) {
-    this.currentQuestion = 1;
+    // First question shall be 0 to prevent nullification of the first array item
+    this.currentQuestion = 0;
```
Comment (jon-shipley) [inline pupil-spa/src/app/services/question/question.service.ts]: ‼️ Doesn't work for me.  

I see `this.currentQuestion` as being the question sequence number.  This not the index position of some array.

Comment (GeorgeChatzigiannis) [reply pupil-spa/src/app/services/question/question.service.ts]: @jon-shipley refactored as discussed

**Comment by activemq** [comment]
{“type”:“CheckCompleteRendered”,“clientTimestamp”:“2017-09-07T13:00:04.139Z”} not being stored in DB, thus the failing the test, this is a bug which George will talk with the Dev's about

File: pupil-spa/src/app/check/check.component.ts [hunk 2]
```diff
@@ -130,12 +135,16 @@ export class CheckComponent implements OnInit {
...
         const matches = /^Q(\d+)$/.exec(stateDesc);
+        this.registerInputService.flush();
         this.question = this.questionService.getQuestion(parseInt(matches[ 1 ], 10));
+        this.registerInputService.initialise();
         this.viewState = 'question';
         break;
       }
       case(/^complete$/).test(stateDesc):
         // Show the check complete screen
+        this.registerInputService.flush();
```
Comment (jon-shipley) [inline pupil-spa/src/app/check/check.component.ts]: Shouldn't `->flush()` be called at the very end, before the `CheckComplete` audit event?

File: pupil-spa/src/app/services/question/question.service.mock.ts [hunk 3]
```diff
@@ -14,10 +18,6 @@ export class QuestionServiceMock {
     return new Question(3, 4, 1);
   }
 
-  getNextQuestionNumber() {
-    return 2;
-  }
-
```
Comment (jon-shipley) [inline pupil-spa/src/app/services/question/question.service.mock.ts]: Thanks for getting rib of this cruft. 👍

File: pupil-spa/src/app/services/question/question.service.ts [hunk 4]
```diff
@@ -14,7 +14,7 @@ export class QuestionService {
   protected config: Config;
 
   constructor(protected storageService: StorageService) {
-    this.currentQuestion = 1;
+    this.currentQuestion = 0;
```
Comment (jon-shipley) [inline pupil-spa/src/app/services/question/question.service.ts]: ‼️  Don't like this at all.   Questions are always numbered from 1 .. n 
Please don't change this.

File: pupil-spa/src/app/check/check.component.ts (hunk 2 above)
Comment (GeorgeChatzigiannis) [reply pupil-spa/src/app/check/check.component.ts]: @jon-shipley Not sure I am following, isn't that what is happening currently?

File: pupil-spa/src/app/services/question/question.service.ts (hunk 4 above)
Comment (GeorgeChatzigiannis) [reply pupil-spa/src/app/services/question/question.service.ts]: Thought I reverted that.
@jon-shipley it's updated

File: pupil-spa/src/app/check/check.component.ts (hunk 2 above)
Comment (jon-shipley) [reply pupil-spa/src/app/check/check.component.ts]: Okay - that's fine - didn't realise it's a different service

**Pull request merged**